ament_pclint/ament_pclint/main.py

```diff
@@ -0,0 +1,372 @@
+    if not pclint_bin:
+        print("Could not find pclint executable 'pclp64'", file=sys.stderr)
+        return 1
+
+    # Prepare pclint command
+    base_cmd = [pclint_bin,
+                '-i"%s"' % args.pclint_config_dir[0],
+                '-i"%s"' % os.path.join(os.path.dirname(os.path.realpath(__file__)), "config"),
+                'env-xml.lnt']
+
+    # Add include directories from arguments
+    if args.include_directories is not None:
+        for directory in args.include_directories:
+            base_cmd.extend(['-i"%s"' % directory])
+
+    # Add compiler definitions from arguments
+    if args.compiler_definitions is not None:
+        for definition in args.compiler_definitions:
+            base_cmd.extend(['-d"%s"' % definition])
+
+    # try to determine the number of CPU cores
+    jobs = 1
+    try:
+        jobs = multiprocessing.cpu_count()
+    except NotImplementedError:
+        # the number of cores cannot be determined, don't parallelize
+        pass
+
+    result = ['' for f in files]
+    idx = 0
+    threads = []
+    retvals = [0 for i in files]
+    with concurrent.futures.ThreadPoolExecutor(max_workers=jobs) as executor:
+        for f in files:
+            retvals[idx] = 0
+            cmd = base_cmd[:]
+            pclint_config_file = get_lnt_file(f, args.pclint_config_file, args.language)
+
+            cmd.extend([pclint_config_file])
+
+            # Add file at the end
+            cmd.append(f)
+
+            # For debug: prints full command to be copied and pasted.
+            if args.debug:
+                print(str(cmd).replace("'", "").replace(",", "").replace("[", "").replace("]", ""))
+
+            # We start one thread per file
+            threads.append(executor.submit(execute_analysis, cmd, result, retvals, idx))
+            idx += 1
+
+    concurrent.futures.wait(threads)
+
+    aggregate_xml = _format_pclint_xml_out(result)
+
+    try:
+        root = ElementTree.fromstring(aggregate_xml)
+    except ElementTree.ParseError as e:
+        print('Invalid XML in pclint output: %s' % str(e),
+              file=sys.stderr)
+        return 1
+
+    # output errors
+    report = {}
+    for doc in root.findall('doc'):
+        for error in doc.findall('error'):
+            report[error.find('file').text] = []
+        for error in doc.findall('error'):
+            filename = error.find('file').text
+            data = {
+                'line': int(error.find('line').text),
+                'id': error.find('code').text,
+                'severity': error.find('type').text,
+                'msg': error.find('desc').text,
+            }
+            report[filename].append(data)
+
+            data = dict(data)
+            data['filename'] = filename
+            print('[%(filename)s:%(line)d]: (%(severity)s: %(id)s) %(msg)s' % data,
+                  file=sys.stderr)
+
+    # output summary
+    error_count = sum(len(r) for r in report.values())
+    error_count += sum(retvals)
```

Comment:
What is the semantic of each `retval` here? Does it really return the number of errors? That might be problematic when that return code wraps depending on the range or is even negative in case of signals?